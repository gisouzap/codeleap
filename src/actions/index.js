import { createSlice } from '@reduxjs/toolkit';

const userSlice = createSlice({
  name: 'user',
  initialState: {
    username: '',
  },

  reducers: {
    saveUsername: (state, action) => {
      state.username = action.payload;
    },
    login: state => {
      console.log(state);
      state.isLoggedIn = true;
    },
    logout: state => {
      state.isLoggedIn = false;
      state.username = '';
    },
  },
});

const postsSlice = createSlice({
  name: 'posts',
  initialState: {},

  reducers: {
    loadPosts: (state, action) => {
      state.items = action.payload;
    },
  },
});

export const { saveUsername, login, logout } = userSlice.actions;
export const { loadPosts, requestPosts } = postsSlice.actions;

export const { reducer: usersReducer } = userSlice;
export const { reducer: postsReducer } = postsSlice;
