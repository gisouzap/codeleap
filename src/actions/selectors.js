const selectUsername = state => state.user.username;

const selectIsLoggedIn = state => state.user.isLoggedIn;

const selectPosts = state => state.posts.items;

export { selectIsLoggedIn, selectUsername, selectPosts };
