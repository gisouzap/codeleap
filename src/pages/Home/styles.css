.home {
  display: flex;
  flex-direction: column;
  align-items: center;
  background-color: var(--color-grey-100);
  width: 100%;
  height: 100vh;
}

.home-wrapper {
  padding: 24px;
  background-color: var(--color-white);
  width: 50vw;
  gap: 16px;
  display: flex;
  height: 100%;
  flex-direction: column;
}

.home-header {
  width: 100%;
  padding: 27px 37px;
  background-color: var(--color-primary);
  display: flex;
  flex-direction: row;
  justify-content: space-between;
}

.home-header p {
  color: var(--color-white);
}

.button-logout {
  background: none;
  border: none;
  color: var(--color-white);
  cursor: pointer;
  transition: 150ms ease-in;
}

.button-logout:hover {
  opacity: 0.8;
}

.form {
  display: flex;
  flex-direction: column;
  gap: 24px;
}

.textarea {
  background: #ffffff;
  border: 1px solid var(--color-grey-300);
  border-radius: 8px;
  padding: 8px 12px;
  width: 100%;
}

.textarea:focus-visible {
  outline-width: 1px;
  outline-color: var(--color-primary);
}

.posts {
  overflow: hidden;
  overflow-y: auto;
  display: flex;
  flex-direction: column;
  gap: 24px;
}

.button-create {
  display: flex;
  justify-content: end;
}
