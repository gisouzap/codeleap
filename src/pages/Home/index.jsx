import { useEffect, useState } from 'react';
import { createPortal } from 'react-dom';
import { useDispatch, useSelector } from 'react-redux';
import { toast } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.css';

import axios from 'axios';

import apiEndpoint from '../../../apiEndpoint';

import Button from '../../components/Button';
import Card from '../../components/Card';
import TextField from '../../components/TextField';
import Post from '../../components/Post';
import Typography from '../../components/Typography';
import DeleteModal from '../../components/DeleteModal';
import EditModal from '../../components/EditModal';

import { loadPosts, logout } from '../../actions';
import {
  selectPosts,
  selectUsername,
  selectIsLoggedIn,
} from '../../actions/selectors';

import './styles.css';

function Home() {
  const dispatch = useDispatch();
  const posts = useSelector(selectPosts);
  const username = useSelector(selectUsername);

  const blankPost = {
    username,
    title: '',
    content: '',
  };

  const [items, setItems] = useState([]);
  const [post, setPost] = useState(blankPost);
  const [currentPost, setCurrentPost] = useState();
  const [showDeleteModal, setShowDeleteModal] = useState(false);
  const [showEditModal, setShowEditModal] = useState(false);
  const [nextPage, setNextPage] = useState();

  const handleLogout = () => dispatch(logout());
  const handleCloseEditModal = () => setShowEditModal(false);
  const handleCloseRemoveModal = () => setShowDeleteModal(false);

  const handleDelete = item => {
    setCurrentPost(item);
    setShowDeleteModal(true);
  };

  const handleEdit = item => {
    setCurrentPost(item);
    setShowEditModal(true);
  };

  const deletePost = async () => {
    try {
      await axios.delete(`${apiEndpoint}${currentPost.id}/`);
      await getPosts();
    } catch (error) {
      console.error(error);
    }

    handleCloseRemoveModal();
    toast.success('Post removed with success!');
  };

  const editPost = async () => {
    try {
      await axios.patch(`${apiEndpoint}${currentPost.id}/`, currentPost);
      await getPosts();
    } catch (error) {
      console.error(error);
    }

    handleCloseEditModal();
    toast.success('Post edited with success!');
  };

  const getPosts = async () => {
    try {
      const response = await axios.get(apiEndpoint);
      dispatch(loadPosts(response.data));
    } catch (error) {
      console.error(error);
    }
  };

  const createPost = async () => {
    try {
      await axios.post(apiEndpoint, post);
      await getPosts();
    } catch (error) {
      console.error(error);
      toast.success('Error!');
    }

    setPost(blankPost);
    toast.success('Post created!');
  };

  const getNextPosts = async () => {
    try {
      const response = await axios.get(nextPage);
      setItems([...items, ...response.data.results]);
      setNextPage(response.data.next);
    } catch (error) {
      console.error(error);
    }
  };

  const handleScroll = e => {
    const isAtBottom =
      e.target.scrollHeight - e.target.scrollTop === e.target.clientHeight;

    if (isAtBottom) {
      getNextPosts();
    }
  };

  useEffect(() => {
    getPosts();
  }, []);

  useEffect(() => {
    if (posts) {
      setItems(posts.results);
      setNextPage(posts.next);
    }
  }, [posts]);

  return (
    <div className="home">
      <div className="home-wrapper">
        <div className="home-header">
          <Typography variant="title">CodeLeap Network</Typography>
          <button onClick={handleLogout} className="button-logout">
            <Typography variant="label">Logout</Typography>
          </button>
        </div>

        <Card>
          <div className="form">
            <Typography variant="title">Whatâ€™s on your mind?</Typography>
            <div>
              <Typography variant="label">Title</Typography>
              <TextField
                onChange={e => setPost({ ...post, title: e.target.value })}
                value={post.title}
              />
            </div>
            <div>
              <Typography variant="label">Content</Typography>
              <textarea
                value={post.content}
                onChange={e => setPost({ ...post, content: e.target.value })}
                className="textarea"
              />
            </div>
            <div className="button-create">
              <Button
                label="create"
                variant="primary"
                onClick={createPost}
                disabled={
                  /^\s*$/.test(post.title) || /^\s*$/.test(post.content)
                }
              />
            </div>
          </div>
        </Card>

        <div
          className="posts"
          data-testid="scroll-element"
          onScroll={handleScroll}
        >
          {items?.map(item => (
            <Post
              item={item}
              key={item.id}
              onDelete={() => handleDelete(item)}
              onEdit={() => handleEdit(item)}
              username={username}
            />
          ))}
        </div>
      </div>

      {showDeleteModal &&
        createPortal(
          <DeleteModal
            onClose={() => setShowDeleteModal(false)}
            onDelete={deletePost}
          />,
          document.body
        )}

      {showEditModal &&
        createPortal(
          <EditModal
            onClose={() => setShowEditModal(false)}
            onSave={editPost}
            onChange={value => setCurrentPost({ ...currentPost, ...value })}
            post={currentPost}
          />,
          document.body
        )}
    </div>
  );
}

export default Home;
