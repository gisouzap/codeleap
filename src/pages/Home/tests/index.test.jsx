import React from 'react';
import { render, fireEvent, act, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import axios from 'axios';
import configureStore from 'redux-mock-store';

import * as selectors from '../../../actions/selectors';

import Home from '../index';

describe('Home', () => {
  const originalPosts = selectors.selectPosts;
  const originalUsername = selectors.selectUsername;

  jest.spyOn(axios, 'get').mockResolvedValue({});

  // silences errors in the terminal because of tests that throw errors
  jest.spyOn(console, 'error').mockImplementation(() => {});

  const posts = {
    results: [
      {
        id: 1,
        username: 'codeLeapUser',
        title: '',
        content: '',
        created_datetime: '"2023-05-09T15:42:38.421Z"',
      },
      {
        id: 2,
        username: 'XPTO',
        title: 'Yey',
        content: 'Hello test',
        created_datetime: '"2023-05-09T15:42:38.421Z"',
      },
    ],
  };

  const username = 'codeLeapUser';

  jest.mock('axios', () => ({
    get: jest.fn(),
    post: jest.fn(),
  }));

  const store = configureStore()({
    user: {
      login: { isLoggedIn: true },
      saveUsername: { username: 'codeLeapUser' },
    },
    posts: {
      loadPosts: {
        items: posts,
      },
    },
  });

  afterEach(() => {
    selectors.selectPosts = originalPosts;
    selectors.selectUsername = originalUsername;

    jest.clearAllMocks();
  });

  afterAll(() => {
    jest.restoreAllMocks();
    jest.clearAllMocks();
  });

  it('should not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');

    render(
      <Provider store={store}>
        <Home />
      </Provider>
    );
    expect(spy).not.toHaveBeenCalled();
  });

  it('should trigger click on logout button', () => {
    const spy = jest.spyOn(global.console, 'error');

    const { getByText } = render(
      <Provider store={store}>
        <Home />
      </Provider>
    );

    const logoutBtn = getByText('Logout');
    fireEvent.click(logoutBtn);

    expect(spy).not.toHaveBeenCalled();
  });

  it('should test the catch if GET is failed', () => {
    jest
      .spyOn(axios, 'get')
      .mockResolvedValue({})
      .mockRejectedValue(new Error('Error GET'));

    render(
      <Provider store={store}>
        <Home />
      </Provider>
    );
  });

  it('should create a new post with title and content', async () => {
    jest.spyOn(axios, 'post').mockResolvedValue({});

    const { getAllByRole, getByText, container } = render(
      <Provider store={store}>
        <Home />
      </Provider>
    );

    const input = getAllByRole('textbox')[0];
    fireEvent.change(input, { target: { value: 'New title' } });

    const textarea = getAllByRole('textbox')[1];
    fireEvent.change(textarea, { target: { value: 'LoL' } });

    const createBtn = getByText('create');

    await act(async () => {
      fireEvent.click(createBtn);
    });
  });

  it('should throw error and show toast error message', async () => {
    jest
      .spyOn(axios, 'post')
      .mockResolvedValue({})
      .mockRejectedValue(new Error('Error POST'));

    const { getAllByRole, getByText, container } = render(
      <Provider store={store}>
        <Home />
      </Provider>
    );

    const input = getAllByRole('textbox')[0];
    fireEvent.change(input, { target: { value: 'New title' } });

    const textarea = getAllByRole('textbox')[1];
    fireEvent.change(textarea, { target: { value: 'LoL' } });

    const createBtn = getByText('create');

    await act(async () => {
      fireEvent.click(createBtn);
    });
  });

  it('should edit an existing post', async () => {
    jest.spyOn(axios, 'patch').mockResolvedValue({});

    selectors.selectPosts = jest.fn(() => posts);
    selectors.selectUsername = jest.fn(() => username);

    const { getAllByRole, getByText, getByTitle } = render(
      <Provider store={store}>
        <Home />
      </Provider>
    );

    const editBtn = getByTitle('Edit');
    fireEvent.click(editBtn);

    const input = getAllByRole('textbox')[2];
    fireEvent.change(input, { target: { value: 'Edited title' } });

    const textarea = getAllByRole('textbox')[3];
    fireEvent.change(textarea, { target: { value: ':)' } });

    const createBtn = getByText('Save');

    await waitFor(() => {
      fireEvent.click(createBtn);
    });
  });

  it('should throw error when try to edit an existing post', async () => {
    jest.spyOn(axios, 'patch').mockResolvedValue({});

    jest
      .spyOn(axios, 'patch')
      .mockResolvedValue({})
      .mockRejectedValue(new Error('Error PATCH'));

    selectors.selectPosts = jest.fn(() => posts);
    selectors.selectUsername = jest.fn(() => username);

    const { getAllByRole, getByText, getByTitle } = render(
      <Provider store={store}>
        <Home />
      </Provider>
    );

    const editBtn = getByTitle('Edit');
    fireEvent.click(editBtn);

    const input = getAllByRole('textbox')[2];
    fireEvent.change(input, { target: { value: 'Edited title' } });

    const textarea = getAllByRole('textbox')[3];
    fireEvent.change(textarea, { target: { value: ':)' } });

    const createBtn = getByText('Save');

    await waitFor(() => {
      fireEvent.click(createBtn);
    });
  });

  it('should delete an existing post', async () => {
    jest.spyOn(axios, 'delete').mockResolvedValue({});

    selectors.selectPosts = jest.fn(() => posts);
    selectors.selectUsername = jest.fn(() => username);

    const { getByText, getByTitle } = render(
      <Provider store={store}>
        <Home />
      </Provider>
    );

    const deleteBtn = getByTitle('Delete');
    fireEvent.click(deleteBtn);

    const confirmBtn = getByText('Delete');

    await waitFor(() => {
      fireEvent.click(confirmBtn);
    });
  });

  it('should throw error when try to delete an existing post', async () => {
    jest
      .spyOn(axios, 'delete')
      .mockResolvedValue({})
      .mockRejectedValue(new Error('Error DELETE'));

    selectors.selectPosts = jest.fn(() => posts);
    selectors.selectUsername = jest.fn(() => username);

    const { getByText, getByTitle } = render(
      <Provider store={store}>
        <Home />
      </Provider>
    );

    const deleteBtn = getByTitle('Delete');
    fireEvent.click(deleteBtn);

    const confirmBtn = getByText('Delete');

    await waitFor(() => {
      fireEvent.click(confirmBtn);
    });
  });

  it('should load more posts on scroll', async () => {
    const nextPage = 'https://dev.codeleap.co.uk/careers/next-page/';

    axios.get.mockResolvedValue({
      data: {
        results: [],
        next: nextPage,
      },
    });

    const { getByTestId } = render(
      <Provider store={store}>
        <Home />
      </Provider>
    );

    const postsDiv = getByTestId('scroll-element');
    Object.defineProperty(postsDiv, 'scrollHeight', { value: 100 });
    Object.defineProperty(postsDiv, 'scrollTop', { value: 50 });
    Object.defineProperty(postsDiv, 'clientHeight', { value: 50 });

    fireEvent.scroll(postsDiv);

    await waitFor(() => {
      expect(axios.get).toHaveBeenCalledTimes(2);
    });
  });

  it('should throw error when try to load more posts on scroll', async () => {
    jest
      .spyOn(axios, 'get')
      .mockResolvedValue({})
      .mockRejectedValue(new Error('Error GET'));

    const { getByTestId } = render(
      <Provider store={store}>
        <Home />
      </Provider>
    );

    const postsDiv = getByTestId('scroll-element');
    Object.defineProperty(postsDiv, 'scrollHeight', { value: 100 });
    Object.defineProperty(postsDiv, 'scrollTop', { value: 50 });
    Object.defineProperty(postsDiv, 'clientHeight', { value: 50 });

    fireEvent.scroll(postsDiv);

    await waitFor(() => {
      expect(axios.get).toHaveBeenCalled();
    });
  });

  it('should scroll posts and not call the request if it is not at the bottom', async () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Home />
      </Provider>
    );

    const postsDiv = getByTestId('scroll-element');
    Object.defineProperty(postsDiv, 'scrollHeight', { value: 100 });
    Object.defineProperty(postsDiv, 'scrollTop', { value: 20 });
    Object.defineProperty(postsDiv, 'clientHeight', { value: 50 });

    fireEvent.scroll(postsDiv);

    await waitFor(() => {
      expect(axios.get).toHaveBeenCalledTimes(1);
    });
  });

  it('should open the edit modal and close it', async () => {
    selectors.selectPosts = jest.fn(() => posts);
    selectors.selectUsername = jest.fn(() => username);

    const { getByText, getByTitle } = render(
      <Provider store={store}>
        <Home />
      </Provider>
    );

    const editBtn = getByTitle('Edit');
    fireEvent.click(editBtn);

    const editTitle = getByText('Edit item');
    expect(editTitle).toBeInTheDocument();

    const cancelBtn = getByText('Cancel');

    fireEvent.click(cancelBtn);

    expect(editTitle).not.toBeInTheDocument();
  });

  it('should open the delete modal and close it', async () => {
    selectors.selectPosts = jest.fn(() => posts);
    selectors.selectUsername = jest.fn(() => username);

    const { getByText, getByTitle } = render(
      <Provider store={store}>
        <Home />
      </Provider>
    );

    const editBtn = getByTitle('Delete');
    fireEvent.click(editBtn);

    const deleteTitle = getByText('Are you sure you want to delete this item?');
    expect(deleteTitle).toBeInTheDocument();

    const cancelBtn = getByText('Cancel');
    fireEvent.click(cancelBtn);

    expect(deleteTitle).not.toBeInTheDocument();
  });

  it('should render and match the snapshot', () => {
    const { container } = render(
      <Provider store={store}>
        <Home />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });
});
