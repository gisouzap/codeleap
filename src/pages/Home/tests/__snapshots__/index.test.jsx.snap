// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should render and match the snapshot 1`] = `
<div>
  <div
    class="home"
  >
    <div
      class="home-wrapper"
    >
      <div
        class="home-header"
      >
        <p
          class="title"
        >
          CodeLeap Network
        </p>
        <button
          class="button-logout"
        >
          <p
            class="label"
          >
            Logout
          </p>
        </button>
      </div>
      <div
        class="card"
      >
        <div
          class="form"
        >
          <p
            class="title"
          >
            Whatâ€™s on your mind?
          </p>
          <div>
            <p
              class="label"
            >
              Title
            </p>
            <input
              class="input"
              type="text"
              value=""
            />
          </div>
          <div>
            <p
              class="label"
            >
              Content
            </p>
            <textarea
              class="textarea"
            />
          </div>
          <div
            class="button-create"
          >
            <button
              class="button primary"
              disabled=""
              type="button"
            >
              create
            </button>
          </div>
        </div>
      </div>
      <div
        class="posts"
        data-testid="scroll-element"
      />
    </div>
  </div>
</div>
`;
